File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        // First I'll look for the correct row and then for the correct column
        int m = matrix.size(), n = matrix[0].size();
        
        int i = 0, j = m - 1;
        while (i < j - 1) {
            int k = (i + j) / 2;
            
            if (matrix[k][0] > target) j = k;
            else i = k;
        }
        
        int row = target >= matrix[j][0] ? j : i;
        i = 0;
        j = n - 1;
        
        while (i <= j) {
            int k = (i + j) / 2;
            
            if (matrix[row][k] == target) return true;
            else if (matrix[row][k] < target) i = k + 1;
            else j = k - 1;
        }
        
        return false;
    }
};

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        low, high = 0, len(matrix)-1
        while low < high - 1:
            mid = (high + low) // 2
            if matrix[mid][0] <= target:
                low = mid
            else:
                high = mid
        row = low
        if matrix[high][0] <= target:
            row = high
        low, high = 0, len(matrix[0])-1
        while low <= high:
            mid = (high + low) // 2
            if matrix[row][mid] == target:
                return True
            if matrix[row][mid] < target:
                low = mid + 1
            else:
                high = mid - 1
        return False
            